performance/include/traccc/performance/timer.hpp

```diff
@@ -0,0 +1,82 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s).
+#include <chrono>
+
+// System include(s).
+#include <memory>
+#include <string>
+
+namespace traccc::performance {
+
+/// Object used for measuring execution time.
+/// Start time measured at construct time. End time at stop().
+/// Creates "list" of times in printout string with names + times.
+/// Creating more than one instance of timer with the same name for the same printout string will lead to incrementing the totalTime for that name
+///
+class timer {
+
+    public:
+
+    /// Constructor when to start time measurement
+    /// @param timer_name name to be printed out identifying what is measured
+    /// @param printout_ptr shared pointer to string where execution times are stored
+    timer(const std::string timer_name, std::shared_ptr<std::string> printout_ptr) : name(timer_name), printout(printout_ptr) {
+        name.resize(nameLength, ' ');
+        
+        // Measure beggining at construct time 
+        beginTime = std::chrono::system_clock::now();
+    }
+
+    /// End time measurement
+    void stop(){
+        // Measure ending at stop() time
+        std::chrono::time_point<std::chrono::system_clock> endTime = std::chrono::system_clock::now();
+        double totalTime = (std::chrono::duration<double>(endTime - beginTime)).count();
+
+        // Look for current name in printout string
+        std::size_t pos = printout->find(name);
+
+        // If this name is not already on the printout string, append it
+        if(pos == std::string::npos){
+            std::string time = std::to_string(totalTime);
+            time.resize(precision, ' ');
+            printout->append(name + time + "\n");
+        }
+        else{
+            // Previous time for this name
+            double previous = std::stof(printout->substr(pos + nameLength));
+            totalTime += previous;
+            std::string time = std::to_string(totalTime);
+            time.resize(precision, ' ');
+            // Replace time in printout string for this name 
+            printout->replace(pos + nameLength, precision, time);
+        }
+        // Release shared_ptr
+        printout.reset();
+    }
+
+    private:
+
+    constexpr static std::size_t precision = 10;
+    constexpr static std::size_t nameLength = 25; 
+
+    /// Name to be printed out from timer
+    std::string name;
+
+    /// Shared pointer to printout string
+    std::shared_ptr<std::string> printout;
```

Comment:
Please don't store this as a pointer to a string. Please either store a file handle so you can write directly to `stderr` or `stdout`, or write the data to a map.

Reply:
May I ask what is wrong with using the pointer to a string? (not doubting that in any way, it's more of a educational question)

Reply:
I initially went with this approach of the shared pointer to allow more than one timer to be running at the same time with the same "printout string" as well as being easy to add the functionality for incrementing same "name" calls together

Reply:
If you consider the two ideas I mentioned before, they each have upsides and downsides; the upside of writing directly to a file handle is immediacy in the sense that the data becomes available to the user immediately. The downside is that the data is unstructured and immalleable: it is (essentially) impossible to retrieve or modify individual points of data. The approach of using a map is the opposite: it is non-immediate (requires dumping it all at the end of the computation, but it is highly structured. The approach you have picked here has the worst of both worlds: it is non-immediate and unstructured, and I personally do not see any upsides to it. The unstructured nature of it requires fragile operations like lookups in a string. Also, this keeps a lot of (shared) state, which is almost always undesirable.

Reply:
Thanks for the explanation, I'll probably take the map route then. Pun intended